src/sources/kubernetes.rs

```diff
@@ -0,0 +1,318 @@
+        if let Some(ValueKind::Bytes(pod_uid)) = log.get(&atom_pod_uid) {
+            // Fetch pod data
+            let data = pod_data.entry(pod_uid.clone()).or_default();
+
+            // Get/determine log format
+            let log_format = if let Some(log_format) = data.log_format {
+                log_format
+            } else {
+                // Detect log format
+                let log_format = if transform_json_message(event.clone()).is_some() {
+                    LogFormat::Json
+                } else if transform_cri_message.transform(event.clone()).is_some() {
+                    LogFormat::CRI
+                } else {
+                    error!(message = "Unknown message format");
+                    // Return untouched message so that user has some options
+                    // in how to deal with it.
+                    LogFormat::Unknown
+                };
+                // Save log format
+                data.log_format = Some(log_format);
+                log_format
+            };
+
+            // Parse message
+            match log_format {
+                LogFormat::Json => transform_json_message(event),
+                LogFormat::CRI => transform_cri_message.transform(event),
+                LogFormat::Unknown => Some(event),
+            }
+        } else {
+            None
+        }
+    })
+}
+
+fn transform_json_message() -> impl FnMut(Event) -> Option<Event> + Send {
+    let mut config = JsonParserConfig::default();
+
+    // Drop so that it's possible to detect if message is in json format
+    config.drop_invalid = true;
+
+    // Drop field
+
+    let mut json_parser: JsonParser = config.into();
+
+    let atom_time = Atom::from("time");
+    let atom_log = Atom::from("log");
+    move |event| {
+        let mut event = json_parser.transform(event)?;
+
+        // Rename fields
+        let log = event.as_mut_log();
+
+        // time -> timestamp
+        if let Some(ValueKind::Bytes(timestamp_bytes)) = log.remove(&atom_time) {
+            match DateTime::parse_from_rfc3339(
+                String::from_utf8_lossy(timestamp_bytes.as_ref()).as_ref(),
+            ) {
+                Ok(timestamp) => log.insert_explicit(
+                    event::TIMESTAMP.clone(),
+                    timestamp.with_timezone(&Utc).into(),
+                ),
+                Err(error) => warn!(message="Non rfc3339 timestamp",error=%error),
+            }
+        } else {
+            warn!(message = "Missing field", field = %atom_time);
+        }
+
+        // log -> message
+        if let Some(message) = log.remove(&atom_log) {
+            log.insert_explicit(event::MESSAGE.clone(), message);
+        } else {
+            warn!(message = "Missing field", field = %atom_log);
+        }
+
+        Some(event)
+    }
+}
+
+fn transform_cri_message() -> Result<Box<dyn Transform>, String> {
+    let mut rp_config = RegexParserConfig::default();
+    // message field
+    rp_config.regex =
```

Comment:
can we make sure we have tests for these?

Reply:
I'll add a test like the `fn file_path_transform()`. Other than that, it is in the execution path of the full tests when using non Docker runtime, like containerd in microk8. Which would fail if this fails.